File: KMP.cpp
#include <iostream>
#include <vector>
#include <string>

using namespace std;

vector<int> BuildNext(const string& pattern) 
{
    int prefix_len = 0; // Length of the previous longest prefix suffix
    vector<int> next(pattern.size(), 0);
    int i = 1;
    
    while (i < pattern.size()) {
        if (pattern[i] == pattern[prefix_len]) 
        {
            prefix_len++;
            next[i] = prefix_len;
            i++;
        } 
        else 
        {
            if (prefix_len != 0) prefix_len = next[prefix_len - 1];
            else 
            {
                next[i] = 0;
                i++;
            }
        }
    }
    return next;
}

int KMPSearch(const string& text, const string& pattern) {
    vector<int>*next = new vector<int>(BuildNext(pattern));
    int i = 0; // Index for text
    int j = 0; // Index for pattern
    
    while (i < text.size()) {
        if (pattern[j] == text[i]) {
            i++;
            j++;
            
            if (j == pattern.size()) 
            {
                delete next;
                return i - j; // Pattern found at index i - j
            }
        } else {
            if (j != 0) {
                j = (*next)[j - 1];
            } else {
                i++;
            }
        }
    }
    delete next;
    return -1; // Pattern not found
}

int main() {
    string text = "ABABDABACDABABCABAB";
    string pattern = "ABABCABAB";
    int result = KMPSearch(text, pattern);
    if (result != -1) {
        cout << "Pattern found at index " << result << endl;
    } else {
        cout << "Pattern not found in the text" << endl;
    }
    system("pause");
    return 0;
}

File: KMP.py
"""
KMP search
"""

def BuildNext(patt):
    """
    Build the longest proper prefix which is also a suffix (lps) array for the pattern.
    """
    next = [0]
    prefix_len = 0 # Length of the previous longest prefix suffix
    i = 1
    while i < len(patt):
        if patt[prefix_len] == patt[i]:
            prefix_len += 1
            next.append(prefix_len)
            i += 1
        else:
            if prefix_len == 0:
                next.append(0)
                i += 1
            else:
                prefix_len = next[prefix_len - 1]
    return next

def KMPSearch(string, patt):
    """
    Search for occurrences of pattern in text using the Knuth-Morris-Pratt (KMP) algorithm.
    """
    next = BuildNext(patt)
    print(next)

    i = 0 # Index for text
    j = 0 # Index for pattern 
    while i < len(string):
        if string[i] == patt[j]:
            i += 1
            j += 1
        elif j > 0:
            j = next[j - 1]
        else:
            i += 1
        if j == len(patt):
            return i - j # Pattern found at index i - j
        
    return -1 # Pattern not found

def Test():
    # Example usage:
    text = "ABABDABACDABABCABAB"
    pattern = "ABABCABAB"
    result = KMPSearch(text, pattern)
    if result != -1:
        print(f"Pattern found at index {result}")
    else:
        print("Pattern not found in the text")

if __name__ == "__main__":
    Test()